### PR TITLE
Add support for publishing execution status updates via a Redis stream.

### PR DESCRIPTION
**Pull request by vadimberezniker**
This will address buildbuddy-io/buildbuddy-internal#590 once it's rolled out & list-based pubsub is disabled.

---

**Version bump**: Minor

### CODE REVIEW

File: enterprise/server/backends/pubsub/pubsub.go [hunk 1]
```diff
@@ -138,8 +138,121 @@ func (p *ListPubSub) Subscribe(ctx context.Context, channelName string) interfac
...
+		if startFromTail {
+			msgs, err := p.rdb.XRevRangeN(ctx, channelName, "+", "-", 1).Result()
+			if err != nil {
+				log.Errorf("Unable to retrieve last element of stream %q: %s", channelName, err)
+				return
+			}
+			if len(msgs) == 1 {
+				msg := msgs[0]
+				streamID = msg.ID
+				if ok := deliverMsg(&msg); !ok {
```
Comment (tylerwilliams) [inline enterprise/server/backends/pubsub/pubsub.go:200]: can you shorten this to

if !deliverMsg(&msg) {
 return
}

File: enterprise/server/backends/pubsub/pubsub.go [hunk 2]
```diff
@@ -138,8 +138,121 @@ func (p *ListPubSub) Subscribe(ctx context.Context, channelName string) interfac
...
+			}
+
+			// We are subscribing to a single stream so there should be exactly one response.
+			if len(result) != 1 {
+				log.Errorf("Did not receive exactly one result for channel %q, got %d", channelName, len(result))
+				return
+			}
+
+			for _, msg := range result[0].Messages {
+				if ok := deliverMsg(&msg); !ok {
```
Comment (tylerwilliams) [inline enterprise/server/backends/pubsub/pubsub.go:225]: (same here)

File: enterprise/server/backends/pubsub/pubsub.go [hunk 3]
```diff
@@ -138,8 +138,121 @@ func (p *ListPubSub) Subscribe(ctx context.Context, channelName string) interfac
...
+	}
+}
+
+// SubscribeHead returns a subscription for all previous and future message on the stream.
+func (p *StreamPubSub) SubscribeHead(ctx context.Context, channelName string) interfaces.Subscriber {
+	// Subscribe from the beginning of the stream.
+	return p.subscribe(ctx, channelName, false /*startFromTail=*/)
+}
+
+// SubscribeTail returns a subscription for messages starting from the last message already on the stream, if any.
```
Comment (tylerwilliams) [inline enterprise/server/backends/pubsub/pubsub.go:243]: ubernit, maybe we could rephrase these to make them clearer?

// SubscribeHead returns a subscription for all past, current, and future messages on the stream.
func (p *StreamPubSub) SubscribeHead(ctx context.Context, channelName string) interfaces.Subscriber {}

// SubscribeTail returns a subscription for all current and future messages on the stream.
func (p *StreamPubSub) SubscribeHead(ctx context.Context, channelName string) interfaces.Subscriber {}

**Review by tylerwilliams — APPROVED** [review]
LGTM

This looks really good!

File: enterprise/server/backends/pubsub/pubsub.go (hunk 1 above)
Comment (vadimberezniker) [reply enterprise/server/backends/pubsub/pubsub.go]: Done.

File: enterprise/server/backends/pubsub/pubsub.go (hunk 2 above)
Comment (vadimberezniker) [reply enterprise/server/backends/pubsub/pubsub.go]: Done.

File: enterprise/server/backends/pubsub/pubsub.go (hunk 3 above)
Comment (vadimberezniker) [reply enterprise/server/backends/pubsub/pubsub.go:243]: The term "current" is a bit ambiguous though as it depends on what the data means to the user of the client.
